chainermn/datasets/shuffle_datablock.py

```diff
@@ -0,0 +1,219 @@
+import itertools
+import random
+
+import chainer
+import functools
+import numpy
+
+
+def _increment_send_counts(table, send_rank, recv_rank, n=1):
+    key = (send_rank, recv_rank)
+    table.setdefault(key, 0)
+    table[key] += n
+
+
+def _flatten1(ary):
+    """1-level flatten"""
+    return functools.reduce(lambda a, b: a + b, ary)
```

Comment:
We'd better just `+` of lists because `INPLACE_ADD` is much faster than `ADD` in Python. Being functional is good, but this case is simple enough I guesss. See https://stackoverflow.com/a/54765916